api/nakadi-event-bus-api.yaml

```diff
@@ -606,22 +488,394 @@ paths:
+          description: Client is not authenticated
+          schema:
+            $ref: '#/definitions/Problem'
+        '403':
+          description: Client is not authorized to perform this operation
+          schema:
+            $ref: '#/definitions/Problem'
+        '404':
+          description: Subscription not found
+          schema:
+            $ref: '#/definitions/Problem'
+        '500':
+          description: Server error
+          schema:
+            $ref: '#/definitions/Problem'
+        '503':
+          description: Service (temporarily) unavailable
+          schema:
+            $ref: '#/definitions/Problem'
+
+  /subscriptions/{subscription_id}/cursors:
+    get:
+      tags:
+        - managed-api
+      security:
+        - oauth2: ['nakadi.event_stream.read']
+      description: Exposes the currently committed offsets of a subscription.
+      parameters:
+        - $ref: '#/parameters/SubscriptionId'
+      responses:
+        '200':
+          description: Ok
+          schema:
+            type: array
+            items:
+              $ref: '#/definitions/Cursor'
+        '401':
+          description: Client is not authenticated
+          schema:
+            $ref: '#/definitions/Problem'
+        '403':
+          description: Client is not authorized to perform this operation
+          schema:
+            $ref: '#/definitions/Problem'
+        '404':
+          description: Subscription not found
+          schema:
+            $ref: '#/definitions/Problem'
+        '500':
+          description: Server error
+          schema:
+            $ref: '#/definitions/Problem'
+        '503':
+          description: Service (temporarily) unavailable
+          schema:
+            $ref: '#/definitions/Problem'
+    put:
+      tags:
+        - managed-api
+      security:
+        - oauth2: ['nakadi.event_stream.read']
+      description: |
+        Endpoint for committing offsets of the subscription. The behavior of commit is specific to `commit_mode`
+        parameter of the client stream (please see details in subscription streaming endpoint specification).
+      parameters:
+        - name: subscription_id
+          in: path
+          type: string
+          description: Id of subscription
+          required: true
+        - name: X-Nakadi-StreamId
+          in: header
+          type: string
+          description: |
+            Id of stream which client uses to read events. It is not possible to make a commit for a terminated or
+            none-existing stream. Also the client can't commit something which was not sent to his stream.
+          required: true
+        - name: cursors
+          in: body
+          description: Cursors with offsets to commit.
+          schema:
+            type: array
+            items:
+              $ref: '#/definitions/Cursor'
+          required: true
+      responses:
+        '200':
+          description: Offsets were committed
+        '207':
```

Comment:
> Why?

@tkrop, ok, I'll try and explain.

In previous discussions up to this, it was emphasised that being able to 'rewind' a cursor offset for the subscription API was a problem, such that clients could never be allowed to that. A specific call to allow a rewind option was a suggestion for the future, as there are valid use cases to allow clients to reset the offset position (it's standard stuff in systems like Kinesis and Kafka). Anyway, it's an error condition, which begs the question why return a success code.

Now, if the thinking is that this call is a partial update or upsert, then I can see your point. But 207 still isn't the right thing:

- The request is a PUT. Returning back 2xx for a PUT request that wasn't honoured is dubious at best. Our API guidelines are pretty clear on the matter - you either replace or you don't and 409 is idiomatic for a conflict scenario, which as I said is my understanding on what's going on here. Now if the consensus is the right design here is upsert/partial, then PUT needs to go. PUT fwiw, led me to think this is not intended to be an upsert on first glance (I have feedback questioning the need for put in general, so I'd be sympathetic to that :).  

- The request isn't a batch/multi. 207 is designed for cases where multiple responses are possible (it has a specific response format for that purpose). That doesn't seem to be the case here, looking at it in more detail. It's worth questioning why 207 is being overloaded to signal partial success for a single request.

So, I hope that helps. If the design is to perform a partial update, then I'd suggest POST plus 200 with a response body the client can examine. If there's a strong need to indicate at the response code level whether the update was complete or partial, then 204 for complete and 200 for partial with a response entity, ie plain HTTP, can support that. PATCH is also an option, but I'd take a pass on PATCH as that comes with the responsibility of defining a patchable format for Cursors or mandating clients using [JSON-Patch](https://tools.ietf.org/html/rfc6902). That seems like overkill here.